test/functional/normalization/json_parsing_test.go

```diff
@@ -0,0 +1,177 @@
+              [
+               [
+                [
+                 [
+                  null
+                 ]
+                ]
+               ]
+              ]
+             ]
+            ]
+          ]
+    }
+    `
+)
+
+var _ = Describe("[LogForwarding] Json log parsing", func() {
+	var (
+		framework *functional.FluentdFunctionalFramework
+		clfb      *functional.ClusterLogForwarderBuilder
+		expected  map[string]interface{}
+	)
+	BeforeEach(func() {
+		_ = json.Unmarshal([]byte(Json), &expected)
+		framework = functional.NewFluentdFunctionalFramework()
+		clfb = functional.NewClusterLogForwarderBuilder(framework.Forwarder).
+			FromInput(logging.InputNameApplication).
+			ToFluentForwardOutput()
+	})
+	AfterEach(func() {
+		framework.Cleanup()
+	})
+
+	It("should parse json message into structured", func() {
+		clfb.Forwarder.Spec.Pipelines[0].Parse = "json"
+		Expect(framework.Deploy()).To(BeNil())
+
+		// Log message data
+		jsonMsg := strings.ReplaceAll(Json, "\n", "")
+		expectedMessage := jsonMsg
+		jsonMsg = strings.ReplaceAll(jsonMsg, "\"", "\\\"")
+		timestamp := "2020-11-04T18:13:59.061892+00:00"
+
+		// Write log line as input to fluentd
+		applicationLogLine := fmt.Sprintf("%s stdout F %s", timestamp, jsonMsg)
+		Expect(framework.WriteMessagesToApplicationLog(applicationLogLine, 10)).To(BeNil())
+
+		// Read line from Log Forward output
+		raw, err := framework.ReadApplicationLogsFrom(logging.OutputTypeFluentdForward)
+		Expect(err).To(BeNil(), "Expected no errors reading the logs")
+
+		// Parse log line
+		var logs []types.ApplicationLog
+		err = types.StrictlyParseLogs(raw, &logs)
+		Expect(err).To(BeNil(), "Expected no errors parsing the logs")
+		Expect(logs[0].Structured).To(Equal(expected), "parsed json message not matching")
+		Expect(logs[0].Message).To(Equal(expectedMessage), "received message not matching")
+	})
+	It("should not parse non json message into structured", func() {
+		clfb.Forwarder.Spec.Pipelines[0].Parse = "json"
+		Expect(framework.Deploy()).To(BeNil())
+
+		// Log message data
+		message := "Functional test message"
+		timestamp := "2020-11-04T18:13:59.061892+00:00"
+
+		// Write log line as input to fluentd
+		applicationLogLine := fmt.Sprintf("%s stdout F %s", timestamp, message)
+		Expect(framework.WriteMessagesToApplicationLog(applicationLogLine, 10)).To(BeNil())
+
+		// Read line from Log Forward output
+		raw, err := framework.ReadApplicationLogsFrom(logging.OutputTypeFluentdForward)
+		Expect(err).To(BeNil(), "Expected no errors reading the logs")
+
+		// Parse log line
+		var logs []types.ApplicationLog
+		err = types.StrictlyParseLogs(raw, &logs)
+		Expect(err).To(BeNil(), "Expected no errors parsing the logs")
+		Expect(logs[0].Structured).To(Equal(map[string]interface{}{}), "received non-empty json parsed structure")
+		Expect(logs[0].Message).To(Equal(message), "received message not matching")
+	})
+	XIt("should not parse invalid json message into structured", func() {
```

Comment:
Why disabled?

Reply:
This is disabled because the parser returns `nil` instead of empty `structured` field

Reply:
we need to fix the behavior for invalid json parsing. the choice of parser gives different results.
There are 3 json parsers available https://docs.fluentd.org/parser/json#json_parser

- oj
- yajl
- json

With each parser we get different output of we parse this invalid json `{"key":"v}`

- oj parses into `"structured":"key"`
- yajl parses into `"structured":"key"`
- json parses into `"structured":{}`

each has different performance characteristics perhaps https://github.com/anilmaurya/fast_jsonparser/blob/master/benchmark/json_benchmark.rb#L38-L41

Cc: @alanconway @jcantrill @eranra

Reply:
oj is the default right?  we should see if we can configure it to be strict https://github.com/ohler55/oj/blob/develop/pages/Modes.md  and then follow-up to maybe choose a faster parser

Reply:
Looks great but we do need to handle invalid JSON. I guess we need to dig into the oj modes.  The json parser produces an acceptable result `"structured":{}` but it's worth some effort to get the faster parser as this is critical path for JSON customers. We don't have much room to further annoy them :)

Reply:
The Oj options are set here
https://github.com/fluent/fluentd/blob/v1.7.4/lib/fluent/env.rb#L24
```
  DEFAULT_OJ_OPTIONS = {bigdecimal_load: :float, mode: :compat, use_to_json: true}
```
other options in the `env.rb` can be overridden via environment variables except this one

Reply:
even in master branch it is same

Reply:
We also get logs
```
2021-03-16 08:49:45 +0000 [info]: Oj is not installed, and failing back to Yajl for json parser
```
Thats why in the avove comment we see `Oj` and `Yajl` giving same result

Reply:
The fluentd documentation gives a hint
https://docs.fluentd.org/quickstart/faq#fluentd-warns-oj-is-not-installed-and-falling-back-to-yajl-for-json-parser

Reply:
I have added the `Oj` json parser to fluentd image. With `oj` parser we get empty `structured` field for invalid json.